core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -258,16 +258,22 @@ private static Schema copyRecord(Schema record, List<Schema.Field> newFields) {
 
   private static Schema.Field copyField(Schema.Field field, Schema newSchema, Integer fieldId) {
     Schema newSchemaReordered;
-    // if the newSchema is an optional schema, make sure the NULL option is always the first
-    if (isOptionSchemaWithNonNullFirstOption(newSchema)) {
+    // if the newSchema is an optional schema with no, or null, default value, then make sure the
```

Comment:
How about `if the newSchema is an optional schema or has a default value that is null, then make sure the NULL option is the first`.

Reply:
(optional_no_default || optional_null_default) != (optional || has_default)